File: 01_GOF23/DP_examples/CSharp/01_Creational/Prototype/ShapeCache.cs
using Creational.Prototype;


namespace Creational;

public class ShapeCache {
    public readonly static ShapeCache Instance = new ShapeCache();
    private ShapeCache() { loadCache(); }
    // 一般从配置表读取
    private void loadCache() {
        RegisterPrototype(ShapeKind.Rectangle, new RectangleProto(new Rectangle()));
        RegisterPrototype(ShapeKind.Circle, new CircleProto(new Circle()));
        RegisterPrototype(ShapeKind.Square, new SquarePrototype(new Square()));
    }

    private Dictionary<ShapeKind, ShapePrototype> shapeMap { get; } = new Dictionary<ShapeKind, ShapePrototype>();
    public void RegisterPrototype(ShapeKind key, ShapePrototype shape) {
        if (!shapeMap.ContainsKey(key))
            shapeMap.Add(key, shape);
    }
    public bool RemovePrototype(ShapeKind key) {
        return shapeMap.Remove(key);
    }
    public Shape GetShapeByWiseClone(ShapeKind Key) {
        if (!shapeMap.TryGetValue(Key, out ShapePrototype proto)) {
            Console.WriteLine($"Clone {Key} failed");
            return null;
        }
        return proto.WiseClone() as Shape;
    }
    public Shape GetShapeByDeepClone(ShapeKind Key) {
        if (!shapeMap.TryGetValue(Key, out ShapePrototype proto)) {
            Console.WriteLine($"Clone {Key} failed");
            return null;
        }
        return proto.DeepClone() as Shape;
    }
 
}

File: 01_GOF23/DP_examples/Go/01_Creational/Prototype/ShapeCache.go
package Prototype

type ShapeCache struct {
	shapeMap map[ShapeKind]IShape
}

var instance *ShapeCache

func ShapeCache_GetInstance() *ShapeCache {
	if instance == nil {
		instance = &ShapeCache{
			shapeMap: make(map[ShapeKind]IShape),
		}
		instance.loadCache()
	}
	return instance
}

func (sc *ShapeCache) loadCache() {
	sc.RegisterPrototype(CIRCLE, NewCircle(CIRCLE, "Circle"))
	sc.RegisterPrototype(SQUARE, NewSquare(SQUARE, "Square"))
	sc.RegisterPrototype(RECTANGLE, NewRectangle(RECTANGLE, "Rectangle"))
}

func (sc *ShapeCache) RegisterPrototype(key ShapeKind, shape IShape) {
	if _, ok := sc.shapeMap[key]; !ok {
		sc.shapeMap[key] = shape
	}
}

func (sc *ShapeCache) RemovePrototype(key ShapeKind) {
	_, ok := sc.shapeMap[key]
	if ok {
		delete(sc.shapeMap, key)
	}
}

func (sc *ShapeCache) GetShape(key ShapeKind) IShape {

	if shape, ok := sc.shapeMap[key]; ok {
		return shape.Clone()
	}
	return nil
}
